lib/plugins/aws/package/compile/functions/index.js

```diff
@@ -441,98 +441,125 @@ class AwsCompileFunctions {
         fileHash.setEncoding('base64');
         versionHash.setEncoding('base64');
 
-        // Read the file in chunks and add them to the hash (saves memory and performance)
-        versionCompilationPromise = BbPromise.fromCallback(cb => {
-          const readStream = fs.createReadStream(artifactFilePath);
-
-          readStream
-            .on('data', chunk => {
-              fileHash.write(chunk);
-              versionHash.write(chunk);
-            })
-            .on('close', () => {
-              cb();
-            })
-            .on('error', error => {
-              cb(error);
+        // Include function and layer configuration in version id hash (without the Code part)
+        versionCompilationPromise = this.addFileContentsToHashes(artifactFilePath, [
+          fileHash,
+          versionHash,
```

Comment:
I believe we should generate just one hash (`versionHash`).

Second hash was introduced with https://github.com/serverless/serverless/pull/4510, where function properties were added to hash that's used in resource logical id.

I take, that over there it was assumed that `CodeSha256` should cover just lambda code, hence `filehHash` was secluded to be used for it. 

However issues as https://github.com/serverless/serverless/issues/7822 show us, it's not exactly the case, and I think AWS documentation also backs that https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-version.html#cfn-lambda-version-codesha256

Therefore let's generate just one `hash` and use it in both places as it was initially

/cc @HyperBrain

Reply:
@pwithams have you had a chance to read this suggestion?

Reply:
Yeah sorry I forgot to mention what I'd found.

It appears the CodeSha256 is specifically a sha256 hash of the lambda code. From the docs you linked (emphasis added): "Use this option to **avoid** publishing a version if the function code **has** changed since you last updated it." So if your zip had somehow changed without your input since you hashed it, it would reject it, maybe for security reasons.

It also says that it will "Only publish a version if the hash value matches the value that's specified". So this sounds like AWS calculates their own hash and checks it against the CodeSha256 provided. That would mean CodeSha256 would have to be created in the same way they are (presumably just a sha256 of the zip) and any "customization" would cause an error.

To test this I changed:

```
const versionDigest = fileDigest.read();
const fileDigest = versionDigest;
```
with
```
const versionDigest = versionHash.read();
const fileDigest = versionDigest;
```

and got this error:

```
 An error occurred: FooLambdaVersionZDGKp5DRfhBVfyaWTjvJ2SSY2fAp0zxjFtV1UafJ8 - CodeSHA256 (ZDGKp5DR+fhBVfyaWTjvJ2SSY2fAp0zxjFtV1U+afJ8=) is different from current CodeSHA256 in $LATEST [...] Please try again with the CodeSHA256 in $LATEST [...] Service: AWSLambdaInternal; Status Code: 400; Error Code: InvalidParameterValueException [...]
```

So it appears that AWS is generating their own digest and checking it against CodeSha256. If so, then CodeSha256 can only be a sha256 of the lambda zip. However, this would obviously not change on property or layer changes so that's why there is the second versionId digest.

Does that make sense? The documentation isn't very clear so I could be wrong.

Reply:
> So it appears that AWS is generating their own digest and checking it against CodeSha256. If so, then CodeSha256 can only be a sha256 of the lambda zip

Thanks for investigating that. It explains then why we use two different hashes. 

Still does it mean that probably the only fix for https://github.com/serverless/serverless/issues/7822 is to tweak lambda package, so hash differs, that's quite problematic.

In light of that we should not consider this PR a fix for #7822

Reply:
For #7822 how did you reproduce the issue? I tried recreating the scenario in this test here but it is passing: https://github.com/serverless/serverless/pull/8066/commits/3ad5651fa49e620f251002b537d93c0dfdf9b7a7#diff-587fde9214d0bca41ecd7959b77058d6L313

Reply:
I've deployed this service: https://github.com/medikoo/test-serverless/blob/f861ad5e66ea97fcbf669e962efd8ca943dfb189/serverless.yml and then I've just added `reservedConcurrency: 1` and redeployed

Reply:
Still couldn't seem to reproduce it when running that config file after `npm install serverless` - is there something missing from the test I linked above? In the meantime I agree we can consider this PR a fix for just my original issue.

Reply:
> In the meantime I agree we can consider this PR a fix for just my original issue.

I believe I've coined a real solution for it here: https://github.com/serverless/serverless/issues/7822#issuecomment-680839963

But let's solve it with other PR, it'll be also good for history, as it's slightly different matter and best if addressed with different commit (note each PR ends as single commit in repository)